src/nnet3/nnet-simple-component.h

```diff
@@ -666,6 +666,55 @@ class NoOpComponent: public NonlinearComponent {
   NoOpComponent &operator = (const NoOpComponent &other); // Disallow.
 };
 
+// GradientClipperComponent just duplicates its input, but clips gradients 
+// during backpropagation if they cross a predetermined threshold.
+// This component will be used to prevent gradient explosion problem in
+// recurrent neural networks
+class GradientClipperComponent: public Component {
+ public:
+  GradientClipperComponent(int32 dim, BaseFloat clipping_threshold, bool norm_based_clipping) {
+    Init(dim, clipping_threshold, norm_based_clipping);
+  }
+  GradientClipperComponent(): dim_(0), clipping_threshold_(-1), norm_based_clipping_(false) { }
+  void Init(int32 dim, BaseFloat clipping_threshold, bool norm_based_clipping);
+  virtual std::string Type() const { return "GradientClipperComponent"; }
+  virtual int32 Properties() const {
+    return kSimpleComponent|kLinearInInput|kPropagateInPlace;
```

Comment:
Your backprop code will work fine in-place, so there is no reason not to add |kBackpropInPlace.